UM/Math/Matrix.py

```diff
@@ -140,6 +168,177 @@ def setByRotationAxis(self, angle, direction, point = None):
             M[:3, 3] = point - numpy.dot(R, point)
         self._data = M
 
+    def compose(self, scale = None, shear = None, angles = None, translate = None, perspective = None):
+        """Return transformation matrix from sequence of transformations.
+
+        This is the inverse of the decompose_matrix function.
+
+        Sequence of transformations:
+            scale : vector of 3 scaling factors
+            shear : list of shear factors for x-y, x-z, y-z axes
+            angles : list of Euler angles about static x, y, z axes
+            translate : translation vector along x, y, z axes
+            perspective : perspective partition of matrix
+
+        >>> scale = numpy.random.random(3) - 0.5
+        >>> shear = numpy.random.random(3) - 0.5
+        >>> angles = (numpy.random.random(3) - 0.5) * (2*math.pi)
+        >>> trans = numpy.random.random(3) - 0.5
+        >>> persp = numpy.random.random(4) - 0.5
+        >>> M0 = compose_matrix(scale, shear, angles, trans, persp)
+        >>> result = decompose_matrix(M0)
+        >>> M1 = compose_matrix(*result)
+        >>> is_same_transform(M0, M1)
+        True
+
+        """
+        M = numpy.identity(4)
+        if perspective is not None:
+            P = numpy.identity(4)
+            P[3, :] = perspective.getData()[:4]
+            M = numpy.dot(M, P)
+        if translate is not None:
+            T = numpy.identity(4)
+            T[:3, 3] = translate.getData()[:3]
+            M = numpy.dot(M, T)
+        if angles is not None:
+            R = Matrix()
+            R.setByEuler(angles.x, angles.y, angles.z, 'sxyz')
+            M = numpy.dot(M, R.getData())
+        if shear is not None:
+            Z = numpy.identity(4)
+            Z[1, 2] = shear.x
+            Z[0, 2] = shear.y
+            Z[0, 1] = shear.z
+            M = numpy.dot(M, Z)
+        if scale is not None:
+            S = numpy.identity(4)
+            S[0, 0] = scale.x
+            S[1, 1] = scale.y
+            S[2, 2] = scale.z
+            M = numpy.dot(M, S)
```

Comment:
Usually you'd do **(scale, shear and perspective) * rotation * translation**, but this seems to be in reverse order. Was that for any reason?

Also I'm not quite sure how the perspective matrix looks in this case. It normally needs 6 parameters, right?